mobile: Replace std::thread with Envoy::Thread::PosixThread

fredyw opened the pull request:
This PR replaces the use of `std::thread` with `Envoy::Thread::PosixThread` (backed by `pthread`). The reason for this change is because `std::thread` may throw an exception and when exceptions are disabled, it will crash the program. For some certain code, such as the cert validation, it should not crash the program when a thread cannot be created, but it should return an error instead.

This PR also refactors the POSIX thread wrapper implementation and exposes the APIs via `Envoy::Thread::PosixThreadFactory` and `Envoy::Thread::PosixThread` so that they can be used directly by Envoy Mobile since Envoy Mobile only supports Android and iOS and those OSes support POSIX. The `Envoy::Thread::PosixThread` has additional functions not available in `Envoy::Thread::Thread`, such as `pthreadId()` and `joinable()` to make the migration from `std::thread` to `Envoy::Thread::PosixThread` easier. The `Envoy::PosixThread::pthreadId()` is especially useful for doing a comparison with `Envoy::PosixThreadFactory::currentPthreadId()`.

Risk Level: medium
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: mobile

fredyw commented:
/retest

fredyw commented:
/assign @abeyad @alyssawilk

abeyad commented on mobile/library/common/thread/posix_thread.h:
```diff
@@ -0,0 +1,78 @@
+  /**
+   * Gets the thread ID. This thread ID may not be unique across different processes.
+   */
+  [[nodiscard]] ThreadId threadId();
```
is this the POSIX thread id?  or the system thread id?  Maybe a comment with a link to the man pages would be helpful.

fredyw replied:
Done.

abeyad commented on mobile/test/common/extensions/cert_validator/platform_bridge/platform_bridge_cert_validator_test.cc:
```diff
@@ -389,6 +390,28 @@ TEST_P(PlatformBridgeCertValidatorTest, DeletedWithValidationPending) {
+  std::string hostname = "server1.example.com";
+  bssl::UniquePtr<STACK_OF(X509)> cert_chain = readCertChainFromFile(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/san_dns2_cert.pem"));
+  // envoy_cert_validation_result result = {ENVOY_SUCCESS, 0, NULL};
```
delete this and the next 2 lines?

abeyad commented on mobile/test/common/thread/posix_thread_test.cc:
```diff
@@ -0,0 +1,38 @@
+  auto thread_factory = PosixThreadFactory::create();
+  std::string message;
+  auto thread =
+      PosixThreadFactory::create()->createThread([&]() { message = "Hello from thread"; });
```
do you want to use the thread_factory created on L8 here?

abeyad reviewed (commented):
this is awesome, thanks @fredyw !

abeyad commented:
adding @danzh2010 since she wrote the original code (I think), in case she has any concerns

danzh2010 commented on mobile/library/common/extensions/cert_validator/platform_bridge/platform_bridge_cert_validator.h:
```diff
@@ -1,17 +1,21 @@
 #include "absl/container/flat_hash_map.h"
 #include "library/common/extensions/cert_validator/platform_bridge/c_types.h"
+#include "library/common/macros/macros.h"
+#include "library/common/thread/posix_thread.h"
```
Somehow I couldn't find where this file is https://github.com/envoyproxy/envoy/tree/main/mobile/library/common/thread.

fredyw replied:
That's a new file created in this PR.

fredyw replied:
This is a stale comment.  I made the changes in `source/common/common/posix/thread_impl.cc` instead.

danzh2010 commented on mobile/library/common/thread/posix_thread.cc:
Would it make sense to change source/common/common/posix/thread_impl.cc by using the PosixThreadHandle instead?

fredyw replied:
Done. I decided to make the changes in `source/common/common/posix/thread_impl.cc` instead.

fredyw commented:
> > Similarly, we can't use `Envoy::Thread::Thread` because, it doesn't have a way to check for the successful or failure status of the thread creation.
> > The `Envoy::Thread::PosixThread` wrapper has similar APIs with `std::thread`, such as `PosixThread::threadId()` and `PosixThread::joinable()` to make the migration from `std::thread` to `Envoy::Thread::PosixThread` easier.
> 
> Can you elaborate it? Envoy::Thread::Thread is just an interface and I assume PosixThread is an implementation of it.

I updated the implementation and also updated the PR description. PTAL.

fredyw commented:
This PR is ready for another review. PTAL.

abeyad commented on mobile/library/common/engine_common.h:
```diff
@@ -41,11 +41,11 @@ class EngineCommon {
   Envoy::TerminateHandler log_on_terminate_;
 #endif
-  std::unique_ptr<Envoy::OptionsImplBase> options_;
+  std::shared_ptr<Envoy::OptionsImplBase> options_;
```
why did this need to be changed to shared_ptr?

fredyw replied:
I have a comment about that in `internal_engine.cc`. This is because there's no way to do a move capture for `std::unique_ptr` with `std::function`. The `std::shared_ptr` is a workaround. I tried switching `std::function` to `absl::AnyInvocable` but ran into MSAN, ASAN, and TSAN. So, the safest way is to keep the existing behavior with `std::function`.

```
// This function takes a `std::shared_ptr` instead of `std::unique_ptr` because `std::function` is a
// copy-constructible type, so it's not possible to move capture `std::unique_ptr` with
// `std::function`.
```

abeyad commented on mobile/library/common/engine_common.h:
```diff
@@ -41,11 +41,11 @@ class EngineCommon {
   DefaultListenerHooks default_listener_hooks_;
   ProdComponentFactory prod_component_factory_;
-  std::unique_ptr<StrippedMainBase> base_;
+  std::shared_ptr<StrippedMainBase> base_;
```
ditto, why did this need to be changed to shared_ptr?

fredyw replied:
Same as above.

abeyad commented on mobile/library/common/extensions/cert_validator/platform_bridge/platform_bridge_cert_validator.cc:
```diff
@@ -84,10 +88,19 @@ ValidationResults PlatformBridgeCertValidator::doVerifyCertChain(
-  std::thread::id thread_id = job.validation_thread_.get_id();
+  Event::Dispatcher& dispatcher = job.result_callback_->dispatcher();
+  job.validation_thread_ = thread_factory_->createThread(
+      [&, certs = std::move(certs), host = std::string(host),
```
do we need to capture everything by reference if we are already passing in certs, hosts, etc?  looks like we only need to capture `this` and `dispatcher` in addition, so we could explicitly capture them instead of `&` (since we're already explicitly creating capture variables for everything else)

fredyw replied:
Makes sense. Done.

abeyad commented on mobile/library/common/extensions/cert_validator/platform_bridge/platform_bridge_cert_validator.h:
```diff
@@ -78,19 +84,20 @@ class PlatformBridgeCertValidator : public CertValidator, Logger::Loggable<Logge
-  void onVerificationComplete(std::thread::id thread_id, std::string hostname, bool success,
-                              std::string error_details, uint8_t tls_alert,
+  void onVerificationComplete(const Thread::ThreadId& thread_id, const std::string& hostname,
+                              bool success, const std::string& error_details, uint8_t tls_alert,
```
this is called in a lambda, in a subsequent epoll loop, are the strings and ThreadId we capture by reference guaranteed to be in scope by the time `onVerificationComplete` is executed?

fredyw replied:
This is called inside `postVerifyResultAndCleanUp`, which is already inside the thread routine (`verifyCertChainByPlatform`). So, all those variables will have the same lifetime as the thread routine.

abeyad reviewed (commented):
awesome, thanks @fredyw !

fredyw commented:
/retest

danzh2010 commented on mobile/library/common/extensions/cert_validator/platform_bridge/platform_bridge_cert_validator.h:
```diff
@@ -55,6 +56,11 @@ class PlatformBridgeCertValidator : public CertValidator, Logger::Loggable<Logge
   }
 
 private:
+  GTEST_FRIEND_CLASS(PlatformBridgeCertValidatorTest, ThreadCreationFailed);
```
Why does the test need to be declared friend?

fredyw replied:
It needs to call the private constructor, so that we can use a a mock `PosixThreadFactory`.

alyssawilk reviewed (commented):
Two thoughts for now - mostly trusting Ali on E-M changes
/wait

fredyw commented on source/common/common/posix/thread_impl.cc:
```diff
@@ -31,93 +34,139 @@ int64_t getCurrentThreadId() {
+        reinterpret_cast<void*>(thread_handle));
+    if (rc != 0) {
+      delete thread_handle;
+      if (crash_on_failure) {
```
Doh, after multiple changes. I realized I had that code removed :( Thanks for pointing it out. I added an early `nullptr`. As for `ENVOY_BUG`, I don't think having a `pthread_create` failure is considered an `ENVOY_BUG`. So, it's probably better for the caller to handle it, whether it can result in an `ENVOY_BUG` or something else.

alyssawilk replied:
mmm, I think it generally is but realistically we're the only call site so if we don't think it merits an envoy bug for "this is unexpected and we want to understand more" for E-M I'm fine leaving as-is.  @abeyad for thoughts on that

abeyad replied:
I agree that pthread_create failing to create a thread isn't a bug, but from our point of view in Envoy (Envoy and Envoy Mobile), any thread creation that fails (whether it's memory, permissions, something else) is an unexpected occurrence, so I think recording it w/ ENVOY_BUG for potential further analysis makes sense.

fredyw commented on source/common/common/posix/thread_impl.h:
```diff
@@ -4,19 +4,99 @@
+   * `crash_on_failure` is set to true, this function will crash when the thread
+   * cannot be created; otherwise a `nullptr` will be returned.
+   */
+  virtual PosixThreadPtr createThread(std::function<void()> thread_routine,
```
I initially had the same thought, but then we would need to modify the `Thread` interface, which also means modifying the Windows wrapper implementation in https://github.com/envoyproxy/envoy/blob/f26361e8838a1e7193f59f7d630a005621c38e94/source/common/common/win32/thread_impl.cc. To keep the scope small, I decided to just restrict the changes to POSIX only for now. If in the future, we want to make the changes to the whole `Thread` interface, I can do that separately. What do you think?

alyssawilk replied:
Totally makes sense - let's leave as-is.

alyssawilk reviewed (approved):
Envoy changes LGTM.

abeyad reviewed (approved):
/lgtm mobile

abeyad commented:
added @lizan for cross company review

abeyad commented:
> added @lizan for cross company review

Actually, per the new policy, this doesn't need cross company review: https://github.com/envoyproxy/envoy/pull/30995/files 

So I'm merging
